pkg/btree/bptree.go

```diff
@@ -85,48 +85,49 @@ func (p *TraversalIterator) init() bool {
 
 // incr moves the iterator by delta, returning false if there is no more data
 // delta is taken to be either -1 or 1.
-func (p *TraversalIterator) incr(delta int) bool {
-	// propagate the carryover
-	if len(p.records) == 1 {
-		// we're at the end of the tree, no more data
+func (p *TraversalIterator) incr(i, delta int) bool {
+	if i == len(p.records) {
+		// we can't increment beyond the root
 		return false
 	}
-	for i := 0; i < len(p.records); i++ {
-		p.records[i].index += delta
-		if p.records[i].index < 0 || p.records[i].index > len(p.records[i].node.Keys) {
-			if i == len(p.records)-1 {
-				// we're at the end of the tree, no more data
-				return false
-			}
+	p.records[i].index += delta
+	rolloverLeft := p.records[i].index < 0
+	rolloverRight := p.records[i].index >= p.records[i].node.NumPointers()
+	if rolloverLeft || rolloverRight {
+		// increment the parent node
+		if !p.incr(i+1, delta) {
+			// if we weren't able to, return false
+			return false
+		}
+		// otherwise, update the current node
```

Comment:
Ok, so we recursively travel to the end of the record, then we carry out the rollover logic from root to leaf